### PR TITLE
Fix phospho scoring

### PR DESCRIPTION
**Pull request by PetraGutenbrunner**
Refactoring:
* divided compute into more sub-methods
* changed the access of all methods except compute from public to protected

AScore.cpp:
62: copy hit, to not loose the previous peptide hit settings

65: score set to zero, the new score type "peptide score" has a different meaning. 
    If for any reason it is not possible to calculate the peptide score, the score should be
    0 to avoid a score mixed up.

82-116: calculate also peptide score for peptides without a phospho event, phospho site 
    or where number of possible sites are equal to the number of phospho events

369 - computeSiteDeterminingIons: calculation of the number of different speaks between the theoretical spectra of the two best scoring peptide permutations, respectively. 

363 - computePermutations_: if there is no phospho-event no permutation calculation needs to be done

482 - peakPickingPerWindowsInSpectrum_: instead of starting the spectrum at 0, the lowest m/z value is rounded down to the next hundred value. Because of that windows with no matches at the beginning are not possible any more.
  
AScore_test.cpp:
* add tests for sub-methods 
* test for calculating the AScore of the peptide sequence mentioned in 
  Beausoleil et al. (values of two best scoring peptides and their phospho sites, and peak depth were manually set)

PhosphoScoring.cpp:
* set the property higherScoreBetter to true

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h [hunk 1]
```diff
@@ -80,29 +80,77 @@ class OPENMS_DLLAPI AScore
...
 
+  protected:
+  
+    template <class InputIterator1, class InputIterator2, class OutputIterator>
+    OutputIterator getSpectrumDifference_(InputIterator1 first1, InputIterator1 last1,
+      InputIterator2 first2, InputIterator2 last2, OutputIterator result) const
+    {
+      while (first1!=last1 && first2!=last2)
+      { 
+        if ((roundf(first1->getMZ() * 100) / 100) < (roundf(first2->getMZ() * 100) / 100))
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: couold you add a textual description what this function does?

/ 100 is not needed and you could use round

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h [hunk 2]
```diff
@@ -80,29 +80,77 @@ class OPENMS_DLLAPI AScore
...
+        {
+          ++first2;
+        }
+        else 
+        { 
+          ++first1; 
+          ++first2; 
+        }
+      }
+      return std::copy(first1,last1,result);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: spaces after comma

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 3]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
 
   PeptideHit AScore::compute(PeptideHit & hit, RichPeakSpectrum & real_spectrum, double fragment_mass_tolerance, bool fragment_mass_unit_ppm) const
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: could probably be const PeptideHit &

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 4]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
...
-
-    // create unmodified variant
-    String umps = without_phospho.toUnmodifiedString();
-
-    // count number of potential phospho sites
-    Size number_of_STY = std::count(umps.begin(), umps.end(), 'S') + std::count(umps.begin(), umps.end(), 'T') + std::count(umps.begin(), umps.end(), 'Y');
-
-    if (real_spectrum.empty() || number_of_phosphorylation_events == 1 || number_of_STY == 0)
+    
+    if (real_spectrum.empty())
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: could be moved to line 66

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 5]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
...
+    
+    // calculate peptide score for each possible phospho site permutation
+    vector<vector<double> > peptide_site_scores(calculatePermutationPeptideScores_(th_spectra, windows_top10, fragment_mass_tolerance, fragment_mass_unit_ppm));
+    
+    // rank peptide permutations ascending
+    multimap<double, Size> ranking(rankWeightedPermutationPeptideScores_(peptide_site_scores));
+    
+    phospho.setScore(ranking.rbegin()->first); // initialize score with highest peptide score (aka highest weighted score)
+    phospho.setSequence(AASequence::fromString(th_spectra[ranking.rbegin()->second].getName()));
+    phospho.setMetaValue("Search_engine_sequence", hit.getSequence().toString());
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: better lower caser "search_engine_sequence"

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 6]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
...
+    
+    if (number_of_phosphorylation_events == 0 || number_of_STY == 0 || number_of_STY == number_of_phosphorylation_events)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: can probably be moved up a bit

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 7]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
...
-      phospho.setSequence(AASequence::fromString(th_spectra[ranking.rbegin()->second].getName()));
-    }
-
-    phospho.setCharge(hit.getCharge());
-    phospho.setMetaValue("Search_engine_sequence", hit.getSequence().toString());
-
+    determineHighestScoringPermutations_(peptide_site_scores, phospho_sites, permutations, ranking);
+    
     Int rank = 1;
     for (vector<ProbablePhosphoSites>::iterator s_it = phospho_sites.begin(); s_it < phospho_sites.end(); ++s_it)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: better to use != instead of <

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 8]
```diff
@@ -60,191 +59,106 @@ namespace OpenMS
...
+    String sequence_str = phospho.getSequence().toString();
+    
+    Size number_of_phosphorylation_events = numberOfPhosphoEvents_(sequence_str);
+    AASequence seq_without_phospho = removePhosphositesFromSequence_(sequence_str);
+  
+    // determine all phospho sites
+    vector<Size> sites(getSites_(seq_without_phospho));
+    Size number_of_STY = sites.size();
+    
+    if (number_of_STY < number_of_phosphorylation_events)
```
Comment (hroest) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: `(number_of_STY < number_of_phosphorylation_events)` -> how is this even possible?

File: src/openms/source/ANALYSIS/ID/AScore.cpp [hunk 9]
```diff
@@ -566,9 +399,152 @@ namespace OpenMS
...
+      
+      for (Size as = 0; as < seq.size(); ++as)
+      {
+        if (as == permutations[i][permu])
+        {
+          seq.setModification(as, "Phospho");
+          ++permu;
+        }
+        
+        if (permu == permutations[i].size()) {
```
Comment (hroest) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: can you put the opening bracket `{` on a new line ?

File: src/openms/source/ANALYSIS/ID/AScore.cpp (hunk 8 above)
Comment (PetraGutenbrunner) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: This case shouldn't happen. This check was already there before I made the changes. I will remove that code lines.

File: src/openms/source/ANALYSIS/ID/AScore.cpp (hunk 6 above)
Comment (PetraGutenbrunner) [inline src/openms/source/ANALYSIS/ID/AScore.cpp]: don't think it is possible to move it up in order to calculate the score also for peptides without an phospho event, phospho site or where the number of phospho events equal phospho sites

**Comment by timosachsenberg** [comment]
thanks. Do we have some ground truth data? e.g. from synthetic phospho libraries that we could use to test it?

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h [hunk 10]
```diff
@@ -78,31 +78,81 @@ class OPENMS_DLLAPI AScore
...
+  
+    /// getSpectrumDifference_ works similar as the method std::set_difference (http://en.cppreference.com/w/cpp/algorithm/set_difference). 
+    /// set_difference was reimplemented, because it was necessary to overwrite the compare operator to be able to compare the m/z values.
+    template <class InputIterator1, class InputIterator2, class OutputIterator>
+    OutputIterator getSpectrumDifference_(InputIterator1 first1, InputIterator1 last1,
+      InputIterator2 first2, InputIterator2 last2, OutputIterator result) const
+    {
+      while (first1 != last1 && first2 != last2)
+      { 
+        if (round(first1->getMZ()) < round(first2->getMZ()))
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: I think comparing only nominal masses is to inexact. Consider that we have multiple ion types (e.g. b and y ions). Two unrelated b and y ions could be considered the same peak while still being up to 1 Da apart in distance.

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h [hunk 11]
```diff
@@ -78,31 +78,81 @@ class OPENMS_DLLAPI AScore
...
         @note the original sequence is saved in the PeptideHits as MetaValue Search_engine_sequence.
     */
-    PeptideHit compute(PeptideHit & hit, RichPeakSpectrum &real_spectrum, double fragment_mass_tolerance, bool fragment_mass_unit_ppm) const;
-
+    PeptideHit compute(const PeptideHit & hit, RichPeakSpectrum &real_spectrum, double fragment_mass_tolerance, bool fragment_mass_unit_ppm) const;
+
+  protected:
+  
+    /// getSpectrumDifference_ works similar as the method std::set_difference (http://en.cppreference.com/w/cpp/algorithm/set_difference). 
+    /// set_difference was reimplemented, because it was necessary to overwrite the compare operator to be able to compare the m/z values.
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: a custom compare comparator can be supplied:
```
template< class InputIt1, class InputIt2,
          class OutputIt, class Compare >
OutputIt set_difference( InputIt1 first1, InputIt1 last1,
                         InputIt2 first2, InputIt2 last2,
                         OutputIt d_first, Compare comp );
```

Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: Something like:
```
struct CompareByMZ
{
  bool is_relative_tolerance;
  double tolerance;
  CompareMZ(double tol, bool is_relative_tol): tolerance(tolerance), is_relative_tolerance(is_relative_tolerance) {}

    bool operator() (const RichPeak1D& lhs, const RichPeak1D& rhs) const
    {
       // compare with relative or absolute tolerances ,,,
    }
};
```

**Comment by timosachsenberg** [comment]
Hi Petra, feel free to ask any question if you need any assistance.

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h (hunk 10 above)
Comment (hroest) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: this should be explained in the comments. 
@PetraGutenbrunner did the original implementation have this property or why is this rounded to integers? maybe using a delta and setting it such that the default is equal to the published implementation would be preferrable...

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h (hunk 11 above)
Comment (hendrikweisser) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: I'm not sure if that would work in every case. As far as I know, the comparison operator ("<") for sets etc. has to impose a total ordering on the domain. However, if you compare with tolerances, that isn't the case, which might lead to contradictions.

For example, if "<" compares with tolerances and you have two elements "a" and "b" that are within the tolerance of each other, then "a < b" and "b < a" are both false. If you now consider an element "c" that is within the tolerance of "b", but not "a", then you start to run into problems.

Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: hmm you are right. I was only thinking about the equality operator.

```
For containers supporting multiple occurrences of a value, the difference includes as many occurrences of a given value as in the first range, minus the amount of matching elements in the second, preserving order.

Notice that this is a directional operation - for a symmetrical equivalent, see set_symmetric_difference.

The elements are compared using operator< for the first version, and comp for the second. Two elements, a and b are considered equivalent if (!(a<b) && !(b<a)) or if (!comp(a,b) && !comp(b,a)).
```

File: src/openms/include/OpenMS/ANALYSIS/ID/AScore.h (hunk 10 above)
Comment (PetraGutenbrunner) [inline src/openms/include/OpenMS/ANALYSIS/ID/AScore.h]: Sorry, it was my mistake. It should be rounded to two decimal places. I will correct this.

**Comment by timosachsenberg** [comment]
Thanks for putting so much effort in this. I merge as the error was related to something different.

**Pull request merged**